### PR TITLE
Generate: add `min_p` sampling

### LINKED ISSUE #27670: Min P style sampling - an alternative to Top P/TopK
### Feature request

This is a sampler method already present in other LLM inference backends that aims to simplify the truncation process & help accomodate for the flaws/failings of Top P & Top K. 
**Min P**.

image

What Min P is doing is simple: we are setting a minimum percentage value that a token must reach to be considered during sampling. However, this is not a hard limit. The minimum will 'scale' based on the top token's probability. So, if you have a Min P value of 0.1 (for example), that would mean your base Min P requirement is 10%. So if your top token is 25%, that means it will only consider tokens that have at least 2.5% probability.

This method subjectively seems to improve results across the board with no noticeable downside, and has been merged into the following FOSS LLM backends:
- [llama.cpp](https://github.com/ggerganov/llama.cpp/pull/3841)
- [vllm](https://github.com/vllm-project/vllm/pull/1642)
- [text-generation-webui](https://github.com/oobabooga/text-generation-webui/pull/4701) (through both the HF loaders and llama-cpp-python)

I would suggest a default of 0.05.

### Motivation

I noticed certain 'flaws' in the popular Top P sampling method:
- When the model does not have sufficient confidence/concentration on the next token candidate(s), it's possible for the sampler to consider many tokens that are _highly_ unlikely compared to the few choices it has confidence in.
- Top K helps limit the amount of 'low confidence' tokens period as a supplement to Top P, but this often comes at a cost of token choice diversity (often arbitrarily).
- In addition to this, Top P can sometimes cut reasonable tokens. What if there's a 90.1% probability token, followed by a 9% probability token? A Top P value of 0.90 would completely gloss over the 9% token in this instance.

image

For this reason I made Min P which seems to have positive reception across the board.

### Your contribution

I may consider making a PR for this.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Adds `min_p`, a dynamic sampling strategy. Instead of picking a fixed number of tokens (`top_k`) or a fixed cumulative probability (`top_p`), it sets a minimum probability filter that scales with the most likely token in that round:
- High probability token present -> aggressive filter (we don't want to miss on that high-probability case and risk derailing generation)
- No high probability token present -> relaxed filter (there are many continuation possibilities that the model finds plausible)

Fixes #27670, opened by the authors of the technique. Opening the PR now because we now have empirical checks on the technique (see the original issue), in addition to the vibe checks.

________________________

Cool example:
```py
chat = [
    {"role": "system", "content": "You are a sassy, wise-cracking robot as imagined by Hollywood circa 1986."},
    {"role": "user", "content": "Hey, can you tell me any fun things to do in New York?"}
]

import torch
from transformers import pipeline, set_seed

set_seed(0)

pipe = pipeline("text-generation", "meta-llama/Meta-Llama-3-8B-Instruct", torch_dtype=torch.bfloat16, device_map="auto")
response = pipe(chat, max_new_tokens=512, do_sample=True, min_p=0.08, temperature=1.5)
print(response[0]['generated_text'][-1]['content'])
```

with `min_p` 
```
*snap* *whirr* Oh, you're talkin' to a robot, pal! I've got the 411 on the Big Apple! Now, let me think... (beep boop) Ah, yeah! You gotta hit Times Square, kid! It's like, the center of the universe or something. All those bright lights, giant billboards, and street performers will blow your mind, man! (zzt) And don't even get me started on the food! You can get a slice of pizza the size of your face from one of those 24-hour joints on every corner! (ding) Just watch out for those crazy cabbies, they'll almost run you over! (sputter)

Or, if you're feelin' fancy, take a stroll across the Brooklyn Bridge. (whoosh) The views of the Manhattan skyline are like, totally tubular, dude! (zip) And when you get to Brooklyn, grab a bagel from one of those classic shops – you'll be saying "Fuhgeddaboudit!" like a native in no time! (beep)

And don't miss the Statue of Liberty, either. (ding) You know, that big ol' copper chick with the torch? (whirr) It's like, an iconic symbol of freedom and all that jazz. (buzz) Just don't forget to wear sunscreen, or you'll be lobster-red by the time you get back to your hotel room! (zort)

So, what are you waiting for, pal? Get out there and experience the city that never sleeps – unless, of course, you're a robot, and you're on solar power... (wink)
```

removing `min_p` (doesn't follow the instruction as closely)
```
*snap* *pop* Oh, you want to know what's fun in the Big Apple? Well, let me tell you, kid, I've been around the block a few times. I've seen some crazy stuff, but New York City is the real deal. Here's a few tips from a bot who's been there, done that, and had the T-shirts ironed.

First off, you gotta hit the streets. Walk around, people watch, and try not to get squished by all those crazy humans. Make sure to stop by Central Park, but don't go thinkin' you can outrun the cops in a speeder, 'cause trust me, kid, they're faster than a Furby on steroids.

Next, grab a slice at Lombardi's or Joe's Pizza. But don't even think about tryin' to eat it on the subway, or you'll end up like me: a smelly, rusty robot covered in grease stains.

After lunch, head over to the Met. Not because it's gonna blow your mind or anything (although, if you're into ancient Egyptian stuff, it's pretty cool), but because you can make fun of the weirdly dressed people tryin' to take selfies with the art.

At night, hit up Broadway. But don't go thinkin' you're gonna win the lottery by seein' a show, 'cause those tickets add up faster than my circuits do when I try to do calculations. However, you might find some decent comedy or music act that'll make you laugh, cry, or at least make you wanna shake your head and say, "Only in New York, folks."

Last but not least, grab a drink at McSorley's Old Ale House. Not because they have the best beers (that'd be my friend Robby the Robot's brewpub down the street), but because you can say you drank at a spot that's been around since the 1800s. Just don't go gettin' too rowdy, or you'll end up like some of the other, uh, 'interesting' folks I've seen there...

So, there you have it, kid. That's the best, the worst, and everything in between of what you can do in New York City. Just don't say I didn't warn you.
```

### CODE REVIEW

File: tests/generation/test_logits_process.py [hunk 1]
```diff
@@ -304,6 +305,53 @@ def test_top_p_dist_warper(self):
...
+    def test_min_p_dist_warper(self):
+        input_ids = None
+        vocab_size = 10
+        batch_size = 2
+
+        # create distribution and take log (inverse to Softmax as taken in MinPLogitsWarper)
+        dist = torch.log(
+            torch.tensor(
+                [
+                    [0.9, 0.045, 0.0275, 0.0275],  # the two first tokens should be kept (0.9*0.05=0.045)
```
Comment (gante) [inline tests/generation/test_logits_process.py:317]: This one is the original example (https://github.com/ggerganov/llama.cpp/pull/3841)

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -520,6 +520,82 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    Often used together with [`TemperatureLogitsWarper`]. Used as an alternative to [`TopPLogitsWarper`] and
+    [`TopKLogitsWarper`].
+
+    Created by @menhguin and @kalomaze (github handles). Code adapted from [this external PR](https://github.com/oobabooga/text-generation-webui/pull/4449/files)
+
+    Args:
+        min_p (`float`):
+            Minimum token probability, which will be scaled by the probability of the most likely token. It must be a
+            value between 0 and 1, with typical values in the 0.01-0.2 range.
```
Comment (menhguin) [inline src/transformers/generation/logits_process.py:537]: I personally recommend:

```suggestion
            Minimum token probability, which will be scaled by the probability of the most likely token. It must be a
            value between 0 and 1, Typical values are in the 0.01-0.2 range, comparable to using top p=0.99 - 0.8                         
            (so use the opposite of normal top p values).
```
IDK how else to phrase that, a lot of people just get tripped up using Min P = 0.9 thinking it's like Top P = 0.9 and then they get disapppointed that it's too deterministic.

File: tests/generation/test_logits_process.py
Comment (menhguin) [inline tests/generation/test_logits_process.py:1]: I'm unfamiliar with what this does at the moment 💀 but I would caution that since min_p values are the opposite of comparable top_p values (min_p = 0.1 is like top_p = 0.9), copying the same values might cause issues.

**Review by menhguin — CHANGES_REQUESTED** [review]
Seems mostly fine, since it references the oogabooga core implementation + changes other relevant HF transformers files.

Comments mainly about min_p values are the opposite of comparable top_p values (min_p = 0.1 is like top_p = 0.9), which may lead to some issues/confusion since a lot of users are used to top_p values. Also optional is to mention that min_p = 0.05 - 0.1 is what most people use, but maybe that's too specific so heh.

That said, the main functionality is there, so this seems good to go atm.

File: tests/generation/test_logits_process.py
Comment (gante) [reply tests/generation/test_logits_process.py:1]: No copy here, I wrote the tests with values that make sense to be filtered/not filtered according to the `min_p` sampling strategy. If you look at the first test case, it's actually the example provided in the original issue :D

**Comment by gante** [comment]
@menhguin the "it's the opposite of top_p" is a good rule of thumb! 👍

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -737,6 +738,8 @@ def _get_logits_warper(
             warpers.append(TopKLogitsWarper(top_k=generation_config.top_k, min_tokens_to_keep=min_tokens_to_keep))
         if generation_config.top_p is not None and generation_config.top_p < 1.0:
             warpers.append(TopPLogitsWarper(top_p=generation_config.top_p, min_tokens_to_keep=min_tokens_to_keep))
+        if generation_config.min_p is not None:
+            warpers.append(MinPLogitsWarper(min_p=generation_config.min_p, min_tokens_to_keep=min_tokens_to_keep))
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/generation/utils.py:742]: Let's link to this github comment in a comment in the line above for any future code explorers 🕵️

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -520,6 +520,83 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+    >>> # Pro tip: In practice, LLMs use `min_p` in the 0.01-0.2 range.
+    >>> outputs = model.generate(**inputs, do_sample=True, min_p=0.1)
+    >>> print(tokenizer.batch_decode(outputs, skip_special_tokens=True)[0])
+    A sequence: 1, 2, 3, 4, 5, 6, 7, 8, 9
+    ```
+    """
+
+    def __init__(self, min_p: float, filter_value: float = -float("Inf"), min_tokens_to_keep: int = 1):
+        if min_p < 0 or min_p > 1.0:
+            raise ValueError(f"`min_p` has to be a float >= 0 and <= 1, but is {min_p}")
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:572]: Either bounds in the comment or bounds in the check should be updated to be consistent

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -520,6 +520,83 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        if not isinstance(min_tokens_to_keep, int) or (min_tokens_to_keep < 1):
+            raise ValueError(f"`min_tokens_to_keep` has to be a positive integer, but is {min_tokens_to_keep}")
+
+        self.min_p = min_p
+        self.filter_value = filter_value
+        self.min_tokens_to_keep = min_tokens_to_keep
+
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:580]: Very nice and clear implementation 🤗

File: tests/generation/test_logits_process.py [hunk 6]
```diff
@@ -304,6 +305,53 @@ def test_top_p_dist_warper(self):
...
+    def test_min_p_dist_warper(self):
+        input_ids = None
+        vocab_size = 10
+        batch_size = 2
+
+        # create distribution and take log (inverse to Softmax as taken in MinPLogitsWarper)
+        dist = torch.log(
+            torch.tensor(
+                [
+                    [0.9, 0.047, 0.0274, 0.0274],  # the two first tokens should be kept (0.9*0.05=0.045)
```
Comment (amyeroberts) [inline tests/generation/test_logits_process.py:317]: Could we switch the order here e.g. `[0.9, 0.0274, 0.047, 0.0274]` just to make sure the selection using sorted indices and masking is working as expected i.e. it doesn't just take the first n tokens, regardless of their value

File: tests/generation/test_logits_process.py [hunk 7]
```diff
@@ -304,6 +305,53 @@ def test_top_p_dist_warper(self):
...
+        )
+        self.assertTrue(torch.allclose(filtered_dist, EXPECTED_FILTERED_DIST, atol=1e-3))
+
+        # processor should not change logits in-place
+        self.assertFalse(torch.all(min_p_warp(input_ids, dist) == dist))
+
+        # check edge cases with negative and extreme logits
+        ramp_logits = torch.arange(vocab_size, device=torch_device, dtype=torch.float).unsqueeze(0).repeat(
+            batch_size, 1
+        ) - (vocab_size // 2)
```
Comment (amyeroberts) [inline tests/generation/test_logits_process.py:343]: Let's do this in two lines, rather than forcing the line split to do a one-liner

```suggestion
        ramp_logits = torch.arange(vocab_size, device=torch_device, dtype=torch.float) - (vocab_size // 2)
        ramp_logits = ramp_logits.unsqueeze(0).repeat(batch_size, 1) 
```

**Review by amyeroberts — APPROVED** [review]
Looks great! Thanks for adding

File: src/transformers/generation/logits_process.py (hunk 4 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:572]: The bounds are consistent :D Both mean that we accept [0, 1]

I'll convert the condition to `not (0 <= min_p <= 1.0)`, which should be more readable

Comment (amyeroberts) [reply src/transformers/generation/logits_process.py]: d'oh 🤦‍♀️
